Index: jdns/jdns.c
===================================================================
--- jdns/jdns.c	(revision 972404)
+++ jdns/jdns.c	(working copy)
@@ -23,6 +23,10 @@
 
 #include "jdns_p.h"
 
+#ifdef JDNS_OS_WIN
+# include <windows.h>
+#endif
+
 #include <time.h>
 
 #include "jdns_packet.h"
Index: jdns/jdns_mdnsd.c
===================================================================
--- jdns/jdns_mdnsd.c	(revision 972404)
+++ jdns/jdns_mdnsd.c	(working copy)
@@ -24,6 +24,10 @@
 
 #include "jdns_mdnsd.h"
 
+#ifdef JDNS_OS_WIN
+# include <windows.h>
+#endif
+
 #include <stdlib.h>
 #include <string.h>
 
Index: jdns/jdns_p.h
===================================================================
--- jdns/jdns_p.h	(revision 972404)
+++ jdns/jdns_p.h	(working copy)
@@ -42,10 +42,6 @@
 # define JDNS_OS_MAC
 #endif
 
-#ifdef JDNS_OS_WIN
-# include <windows.h>
-#endif
-
 #ifdef JDNS_OS_UNIX
 # include <unistd.h>
 # include <netinet/in.h>
Index: jdns/jdns_packet.c
===================================================================
--- jdns/jdns_packet.c	(revision 972404)
+++ jdns/jdns_packet.c	(working copy)
@@ -25,6 +25,10 @@
 
 #include "jdns_p.h"
 
+#ifdef JDNS_OS_WIN
+# include <windows.h>
+#endif
+
 /*  maximum length of a sublabel */
 #define MAX_SUBLABEL_LENGTH  63
 
Index: jdns/jdns_util.c
===================================================================
--- jdns/jdns_util.c	(revision 972404)
+++ jdns/jdns_util.c	(working copy)
@@ -25,6 +25,10 @@
 
 #include "jdns_packet.h"
 
+#ifdef JDNS_OS_WIN
+# include <windows.h>
+#endif
+
 /* ---------------------------------------------------------------------------- */
 /*  misc */
 /* ---------------------------------------------------------------------------- */
Index: xmpp/cutestuff/socks.cpp
===================================================================
--- xmpp/cutestuff/socks.cpp	(revision 972404)
+++ xmpp/cutestuff/socks.cpp	(working copy)
@@ -34,7 +34,7 @@
 
 #ifdef Q_OS_WIN32
 #include <windows.h>
-#include <winsock2.h>
+//#include <winsock2.h>
 #endif
 
 #ifdef Q_OS_UNIX
Index: xmpp/jingle/jingletasks.cpp
===================================================================
--- xmpp/jingle/jingletasks.cpp	(revision 972404)
+++ xmpp/jingle/jingletasks.cpp	(working copy)
@@ -325,8 +325,8 @@
 			}
 			else
 			{
-				QNetworkInterface *interface = new QNetworkInterface();
-				QList<QHostAddress> ips = interface->allAddresses();
+				QNetworkInterface *inter = new QNetworkInterface();
+				QList<QHostAddress> ips = inter->allAddresses();
 				qSort(ips.begin(), ips.end(), interfaceOrder);
 	
 				if (ips.count() == 0)
@@ -529,8 +529,8 @@
 		}
 		else
 		{
-			QNetworkInterface *interface = new QNetworkInterface();
-			QList<QHostAddress> ips = interface->allAddresses();
+			QNetworkInterface *inter = new QNetworkInterface();
+			QList<QHostAddress> ips = inter->allAddresses();
 			qSort(ips.begin(), ips.end(), interfaceOrder);
 
 			if (ips.count() == 0)
Index: xmpp/xmpp-im/s5b.cpp
===================================================================
--- xmpp/xmpp-im/s5b.cpp	(revision 972404)
+++ xmpp/xmpp-im/s5b.cpp	(working copy)
@@ -32,7 +32,7 @@
 
 #ifdef Q_OS_WIN
 # include <windows.h>
-# include <winsock2.h>
+//# include <winsock2.h>
 #else
 # include <netinet/in.h>
 #endif
